File: 283. 移动零.py
class Solution(object):
    def moveZeroes(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        # 未遇到0时快慢指针指向同一个数，一起向右移动；
        # 遇到0时慢指针停下，快指针指向下一个数，交换，慢指针右移，二者再次相等
        slow = 0  
        for fast in range(len(nums)):
            if nums[fast]!=0:
                nums[slow],nums[fast]=nums[fast],nums[slow]
                slow+=1


nums = [0, 1, 0, 3, 12]
s = Solution()
s.moveZeroes(nums)
print(nums)

File: 283. 移动零.go
package main

func moveZeroes(nums []int) {
	r := 0
	for l := range len(nums) {
		if nums[l] != 0 {
			continue
		}
		if r <= l {
			r = l + 1
		}
		for r < len(nums) && nums[r] == 0 {
			r++
		}
		if r == len(nums) {
			return
		}
		nums[l], nums[r] = nums[r], nums[l]
	}
}
